libwnbd/wnbd_ioctl.cpp

```diff
@@ -133,6 +137,315 @@ DWORD WnbdOpenAdapterCMDeviceInstance(PDEVINST DeviceInstance)
+    BOOL Cleanup = FALSE, Found = FALSE;
+    SP_DRVINFO_DATA_A DrvInfoData;
+    DrvInfoData.cbSize = sizeof DrvInfoData;
+    DWORD Status = ERROR_SUCCESS;
+
+    InstallParams.cbSize = sizeof InstallParams;
+    if (!SetupDiGetDeviceInstallParamsA(AdapterDevHandle, DevInfoData, &InstallParams)) {
+        Status = GetLastError();
+        goto exit;
+    }
+
+    InstallParams.FlagsEx |= DI_FLAGSEX_INSTALLEDDRIVER | DI_FLAGSEX_ALLOWEXCLUDEDDRVS;
+    if (!SetupDiSetDeviceInstallParamsA(AdapterDevHandle, DevInfoData, &InstallParams)) {
+        Status = GetLastError();
+        goto exit;
+    }
+
+    if (!SetupDiBuildDriverInfoList(AdapterDevHandle, DevInfoData, SPDIT_CLASSDRIVER)) {
+        Status = GetLastError();
+        goto exit;
+    }
+    Cleanup = TRUE;
+
+    for (DWORD index = 0;
+        SetupDiEnumDriverInfoA(AdapterDevHandle, DevInfoData, SPDIT_CLASSDRIVER, index, &DrvInfoData);
+        index++) {
+
+        if (!SetupDiGetDriverInfoDetailA(AdapterDevHandle, DevInfoData, &DrvInfoData,
+                                         DrvDetailData, DrvDetailData->cbSize, 0)
+            && ERROR_INSUFFICIENT_BUFFER != GetLastError()) {
+            continue;
+        }
+
+        Found = TRUE;
+        break;
+    }
+    if (!Found) {
+        Status = GetLastError();
+        goto exit;
+    }
+
+
+exit:
+    if (ERROR_SUCCESS != Status) {
+        LogError(
+            "Could not get installed driver information data. "
+            "Error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+    }
+
+    if (Cleanup) {
+        SetupDiDestroyDriverInfoList(AdapterDevHandle, DevInfoData, SPDIT_COMPATDRIVER);
+    }
+
+    return Status;
+}
+
+DWORD WnbdRemoveDriver(HDEVINFO AdapterDevHandle, PSP_DEVINFO_DATA DevInfoData, PBOOL RebootRequired)
+{
+    SP_DRVINFO_DETAIL_DATA_A DrvDetailData = { 0 };
+    DrvDetailData.cbSize = sizeof DrvDetailData;
+    DWORD Status = ERROR_SUCCESS;
+
+    Status = GetDrvDetailData(AdapterDevHandle, DevInfoData, &DrvDetailData);
+    if (ERROR_SUCCESS != Status) {
+        return Status;
+    }
+
+    std::string SearchString(DrvDetailData.InfFileName);
+
+    if (!DiUninstallDevice(0, AdapterDevHandle, DevInfoData, 0, RebootRequired)) {
+        Status = GetLastError();
+    }
+
+    if (SearchString.length()) {
+        SetupUninstallOEMInfA(
+            SearchString.substr(SearchString.find_last_of("\\") + 1).c_str(), SUOI_FORCEDELETE, 0);
+    }
+
+    if (ERROR_SUCCESS != Status) {
+        LogError(
+            "Could not remove driver. "
+            "Error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+    }
+    return Status;
+}
+
+
+static DWORD FindDevice(HDEVINFO* AdapterDevHandle, SP_DEVINFO_DATA* DevInfoData)
+{
+    CHAR TempBuf[2048];
+    memset(TempBuf, 0, sizeof(TempBuf));
+    BOOL Found = FALSE;
+    DWORD Status = ERROR_FILE_NOT_FOUND;
+
+    *AdapterDevHandle = SetupDiGetClassDevsA(0, 0, 0, DIGCF_ALLCLASSES | DIGCF_PRESENT);
```

Comment:
Should we use ``WNBD_GUID`` as ``WnbdOpenAdapterEx`` does? We could print a warning when there are multiple WNBD devices.